<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="PS">
<meta name="dcterms.date" content="2024-12-09">

<title>Single-Cell RNA-Seq Analysis with Seurat</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="scRNAseq-course_files/libs/clipboard/clipboard.min.js"></script>
<script src="scRNAseq-course_files/libs/quarto-html/quarto.js"></script>
<script src="scRNAseq-course_files/libs/quarto-html/popper.min.js"></script>
<script src="scRNAseq-course_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="scRNAseq-course_files/libs/quarto-html/anchor.min.js"></script>
<link href="scRNAseq-course_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="scRNAseq-course_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="scRNAseq-course_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="scRNAseq-course_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="scRNAseq-course_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-full">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#introduction-to-single-cell-rna-sequencing-scrna-seq" id="toc-introduction-to-single-cell-rna-sequencing-scrna-seq" class="nav-link active" data-scroll-target="#introduction-to-single-cell-rna-sequencing-scrna-seq">1. Introduction to Single-Cell RNA Sequencing (scRNA-seq)</a></li>
  <li><a href="#setting-up-the-environment" id="toc-setting-up-the-environment" class="nav-link" data-scroll-target="#setting-up-the-environment">2. Setting Up the Environment</a>
  <ul class="collapse">
  <li><a href="#required-packages" id="toc-required-packages" class="nav-link" data-scroll-target="#required-packages">2.1 Required Packages</a></li>
  <li><a href="#loading-packages" id="toc-loading-packages" class="nav-link" data-scroll-target="#loading-packages">2.2 Loading Packages</a></li>
  </ul></li>
  <li><a href="#loading-and-preprocessing-data" id="toc-loading-and-preprocessing-data" class="nav-link" data-scroll-target="#loading-and-preprocessing-data">3. Loading and Preprocessing Data</a>
  <ul class="collapse">
  <li><a href="#loading-data" id="toc-loading-data" class="nav-link" data-scroll-target="#loading-data">3.1 Loading Data</a></li>
  <li><a href="#quality-control" id="toc-quality-control" class="nav-link" data-scroll-target="#quality-control">3.2 Quality Control</a></li>
  <li><a href="#doublet-removal" id="toc-doublet-removal" class="nav-link" data-scroll-target="#doublet-removal">3.3 Doublet Removal</a></li>
  <li><a href="#final-outlier-filtering" id="toc-final-outlier-filtering" class="nav-link" data-scroll-target="#final-outlier-filtering">3.4 Final Outlier Filtering</a></li>
  </ul></li>
  <li><a href="#normalization-and-scaling" id="toc-normalization-and-scaling" class="nav-link" data-scroll-target="#normalization-and-scaling">4. Normalization and Scaling</a>
  <ul class="collapse">
  <li><a href="#normalization" id="toc-normalization" class="nav-link" data-scroll-target="#normalization">4.1 Normalization</a></li>
  <li><a href="#finding-variable-features-scaling" id="toc-finding-variable-features-scaling" class="nav-link" data-scroll-target="#finding-variable-features-scaling">4.2 Finding variable features &amp; Scaling</a></li>
  </ul></li>
  <li><a href="#dimensionality-reduction-and-clustering" id="toc-dimensionality-reduction-and-clustering" class="nav-link" data-scroll-target="#dimensionality-reduction-and-clustering">5. Dimensionality Reduction and Clustering</a>
  <ul class="collapse">
  <li><a href="#principal-component-analysis-pca" id="toc-principal-component-analysis-pca" class="nav-link" data-scroll-target="#principal-component-analysis-pca">5.1 Principal Component Analysis (PCA)</a></li>
  <li><a href="#clustering-and-visualization" id="toc-clustering-and-visualization" class="nav-link" data-scroll-target="#clustering-and-visualization">5.2 Clustering and Visualization</a></li>
  <li><a href="#integration-using-harmony" id="toc-integration-using-harmony" class="nav-link" data-scroll-target="#integration-using-harmony">5.3 Integration Using Harmony</a></li>
  </ul></li>
  <li><a href="#saving-your-processed-object" id="toc-saving-your-processed-object" class="nav-link" data-scroll-target="#saving-your-processed-object">6. Saving your processed object</a></li>
  <li><a href="#annotating-clusters" id="toc-annotating-clusters" class="nav-link" data-scroll-target="#annotating-clusters">7. Annotating Clusters</a></li>
  <li><a href="#extras" id="toc-extras" class="nav-link" data-scroll-target="#extras">8. Extras</a>
  <ul class="collapse">
  <li><a href="#pathway-analysis-with-clusterprofiler" id="toc-pathway-analysis-with-clusterprofiler" class="nav-link" data-scroll-target="#pathway-analysis-with-clusterprofiler">8.1 Pathway Analysis with clusterProfiler</a></li>
  <li><a href="#identifying-number-of-clusters" id="toc-identifying-number-of-clusters" class="nav-link" data-scroll-target="#identifying-number-of-clusters">8.2 Identifying number of clusters</a></li>
  <li><a href="#cell-cycle-analysis" id="toc-cell-cycle-analysis" class="nav-link" data-scroll-target="#cell-cycle-analysis">8.3 Cell Cycle Analysis</a></li>
  <li><a href="#species-classification" id="toc-species-classification" class="nav-link" data-scroll-target="#species-classification">8.4 Species classification</a></li>
  </ul></li>
  </ul>
</nav>
</div>
<main class="content column-page-left" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Single-Cell RNA-Seq Analysis with Seurat</h1>
</div>



<div class="quarto-title-meta column-page-left">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>PS </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">December 9, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="introduction-to-single-cell-rna-sequencing-scrna-seq" class="level1">
<h1>1. Introduction to Single-Cell RNA Sequencing (scRNA-seq)</h1>
<p>A variety of toolkits and analytical frameworks have been developed to facilitate scRNA-seq data analysis. Among the most widely used are Seurat, created by Rahul Satija’s lab for R, and scanpy, developed by Fabian Theis’s lab for Python. Both provide robust functions and extensive parameter sets that cover many of the routine analyses commonly performed on scRNA-seq data. However, it’s essential to recognize that these frameworks may not encompass all possible analyses, so exploring additional tools can broaden the scope of your data insights.</p>
<p>In this tutorial, aimed at beginners, we will primarily focus on using Seurat to analyze scRNA-seq data in R. Throughout this tutorial, we’ll introduce some additional tools that offer complementary functionalities beyond Seurat’s capabilities.</p>
<p>This tutorial assumes that preprocessing steps—such as base calling, mapping, and read counting—have already been completed. For data generated with the 10x Genomics Chromium Single Cell Gene Expression Solution, the Cell Ranger pipeline from 10x Genomics is commonly used, resulting in a count matrix. If your data is from a different technology (e.g., well-based experiments like Smart-Seq2), the Cell Ranger pipeline may not be applicable, and an alternative method will be required to generate the count matrix.</p>
<hr>
</section>
<section id="setting-up-the-environment" class="level1">
<h1>2. Setting Up the Environment</h1>
<section id="required-packages" class="level3">
<h3 class="anchored" data-anchor-id="required-packages">2.1 Required Packages</h3>
<p>Before starting, ensure you have the following packages installed:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Install CRAN packages</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="fu">c</span>(<span class="st">"Seurat"</span>, <span class="st">"dplyr"</span>, <span class="st">"ggplot2"</span>, <span class="st">"robustbase"</span>, <span class="st">"patchwork"</span>,<span class="st">"devtools"</span>, </span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>                   <span class="st">"scRNAseq"</span>, <span class="st">"SingleR"</span>, <span class="st">"tibble"</span>, <span class="st">"viridis"</span>, <span class="st">"clusterProfiler"</span>))</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Install Bioconductor packages</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">"BiocManager"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)) {</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="fu">c</span>(<span class="st">"scater"</span>, <span class="st">"harmony"</span>, <span class="st">"org.Hs.eg.db"</span>))</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Install DoubletFinder from GitHub (since it is not available on CRAN)</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">"chris-mcginnis-ucsf/DoubletFinder"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="loading-packages" class="level3">
<h3 class="anchored" data-anchor-id="loading-packages">2.2 Loading Packages</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load libraries</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(robustbase)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scater)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DoubletFinder)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(harmony)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scRNAseq)</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SingleR)</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tibble)</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(viridis)</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(clusterProfiler)</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(org.Hs.eg.db)</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>sampleColors <span class="ot">&lt;-</span>  <span class="fu">c</span>(<span class="st">"#2A363B"</span>, <span class="st">"#019875"</span>, <span class="st">"#FF847C"</span>, <span class="st">"#E84A5F"</span>)</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>clusterColors <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>  <span class="st">"#C0392B"</span>, <span class="st">"#96281B"</span>, <span class="st">"#B0C4B1"</span>, <span class="st">"#D9A441"</span>, </span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>  <span class="st">"#577284"</span>, <span class="st">"#4A235A"</span>, <span class="st">"#D7BDE2"</span>, <span class="st">"#8C6E63"</span>,</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>  <span class="st">"#1F618D"</span>, <span class="st">"#FF5733"</span>, <span class="st">"#28B463"</span>, <span class="st">"#F39C12"</span>,</span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>  <span class="st">"#8E44AD"</span>, <span class="st">"#16A085"</span>, <span class="st">"#E74C3C"</span></span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
</section>
<section id="loading-and-preprocessing-data" class="level1">
<h1>3. Loading and Preprocessing Data</h1>
<p>In this section, we will load a sample dataset and perform quality control to filter out low-quality and doublet cells.</p>
<section id="loading-data" class="level3">
<h3 class="anchored" data-anchor-id="loading-data">3.1 Loading Data</h3>
<p>We will use four midbrain organoid samples sequenced using 10X scRNA-seq at two timepoints. The data set comes from this <a href="https://www.nature.com/articles/s41467-021-27464-5">publication</a>.</p>
<p>The goal of this section is to load the raw expression data from 10X CellRanger and to create a Seurat object. A Seurat object is organized into multiple slots, including assays (storing expression data), meta.data (holding cell-level annotations), and reductions (storing dimensionality reduction results like PCA or UMAP). This design facilitates seamless preprocessing, visualization, and statistical analysis workflows within a single framework. The flexibility of the Seurat object allows for efficient handling of large datasets and integration of multi-modal data, such as RNA and protein measurements.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># List all directories within the "data/" directory, excluding the top-level "data/" itself.</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>dirs <span class="ot">&lt;-</span> <span class="fu">list.dirs</span>(<span class="st">"data/organoids/"</span>)[<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co"># This step makes sure that the "orig.ident" variable in your Seurat object will be informaticve.</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(dirs) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"day60-1"</span>,<span class="st">"day60-2"</span>, <span class="st">"day30-1"</span>,<span class="st">"day30-2"</span>)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in the 10X Genomics data from each specified directory in 'dirs'.</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="co"># The 'Read10X' function reads the gene expression count data for each sample into one large matrix</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>da_diff.data <span class="ot">&lt;-</span> <span class="fu">Read10X</span>(<span class="at">data.dir =</span> dirs)</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a Seurat object with the count data. This is an essential structure in Seurat for</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="co"># storing and analyzing single-cell RNA-seq data.</span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a><span class="co"># - 'min.cells = 3' filters out genes not expressed in at least 3 cells.</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a><span class="co"># - 'min.features = 200' filters out cells that have fewer than 200 detected genes.</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>da_diff <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> da_diff.data, <span class="at">min.cells =</span> <span class="dv">3</span>, <span class="at">min.features =</span> <span class="dv">200</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="quality-control" class="level3">
<h3 class="anchored" data-anchor-id="quality-control">3.2 Quality Control</h3>
<p>A number of factors should be examined before downstream analyses, many of which we’ll address here:</p>
<ul>
<li><p>Low library size: When cells are very degraded or absent from the library preparation, the number of reads sequenced from that library will be very low. It’s important to remove these cells from downstream analyses.</p></li>
<li><p>Low number of expressed genes: A low number of expressed genes (which is highly correlated with low library size) may be a result of poor-quality cells (e.g.&nbsp;dying, degraded, damaged, etc.), followed by high PCR amplification of the remaining RNA. Again, these cells should be removed from downstream analyses.</p></li>
<li><p>High mitochondrial gene content: High concentrations of mitochondrial genes is often a result of damaged cells where the endogenous RNA escapes or degrades. As mitochondria has its own cell membranes, it is often the last DNA/RNA in damaged cells to degrade and hence occurs in high quantities during sequencing.</p>
<p>Fraction of reads originating from the mitochondria will be heavily dependent on if you’re sequencing cells or nuclei. mtRNA should not be present in nuclei preparations and is usually &lt; 0.1%. For scRNA-seq 5-10% can be considered.</p></li>
<li><p>Batch effect: Large scRNA-seq projects usually need to generate data across multiple batches due to logistical constraints. However, the processing of different batches is often subject to variation, e.g., changes in operator, differences in reagent quality and concentration, the sequencing machine used, etc. This results in systematic differences in the observed expression in cells from different batches, which we refer to as “batch effects”. Batch effects are problematic as they can be major drivers of variation in the data, masking the relevant biological differences and complicating interpretation of the results. We will look into removal of batch effects in section 5.</p></li>
</ul>
<section id="calculating-and-inspecting-basic-qc-paramters" class="level4">
<h4 class="anchored" data-anchor-id="calculating-and-inspecting-basic-qc-paramters">Calculating and inspecting basic QC paramters</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the percentage of mitochondrial genes</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>da_diff[[<span class="st">"percent.mt"</span>]] <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(da_diff, <span class="at">pattern =</span> <span class="st">"^MT-"</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize QC metrics</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of genes per cell ("nFeature_RNA") and UMIs per cell ("nFeature_RNA") was calculated when we intiated the Seurat object</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(da_diff, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"percent.mt"</span>),</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">ncol =</span> <span class="dv">3</span>,<span class="at">pt.size =</span> <span class="dv">0</span>)<span class="sc">+</span><span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">5</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">&amp;</span> <span class="fu">scale_fill_manual</span>(<span class="at">values=</span>sampleColors)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Default search for "data" layer in "RNA" assay yielded no results;
utilizing "counts" layer instead.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="scRNAseq-course_files/figure-html/quality-control-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="automated-identification-of-cutoffs-for-filtering" class="level4">
<h4 class="anchored" data-anchor-id="automated-identification-of-cutoffs-for-filtering">Automated identification of cutoffs for filtering</h4>
<p>Automated outlier detection is a useful method for identifying low-quality cells in scRNA-seq datasets. It removes the need for manual thresholding, which can be subjective and inconsistent, by relying on statistical measures to flag problematic cells. This approach is particularly helpful for datasets with complex or variable distributions, as it adjusts dynamically to the data. It also allows for multidimensional detection, identifying cells with unusual combinations of metrics rather than focusing on individual values.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a matrix of selected statistics from da_diff for analysis</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>stats <span class="ot">&lt;-</span> <span class="fu">cbind</span>(da_diff<span class="sc">$</span>percent.mt, da_diff<span class="sc">$</span>nFeature_RNA, da_diff<span class="sc">$</span>nCount_RNA)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate outlyingness scores for each cell based on the selected statistics</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="co"># we will append tthe result to the meta data of the seurat object</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="co"># only.outlyingness = TRUE returns only the outlyingness values</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>da_diff<span class="sc">$</span>outlying <span class="ot">&lt;-</span> <span class="fu">adjOutlyingness</span>(stats, <span class="at">only.outlyingness =</span> <span class="cn">TRUE</span>)</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify cells that are statistical outliers based on high outlyingness scores</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>outlier <span class="ot">&lt;-</span> <span class="fu">isOutlier</span>(da_diff<span class="sc">$</span>outlying, <span class="at">type =</span> <span class="st">"higher"</span>)</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Summarize the number of outliers identified</span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(outlier)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Mode   FALSE    TRUE 
logical    9208     946 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Add a column to da_diff indicating if each cell is an outlier</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>da_diff<span class="sc">$</span>is.outlier <span class="ot">&lt;-</span> outlier</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot nFeature_RNA against nCount_RNA, coloring points by outlier status</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Facet by 'orig.ident' to visualize each sample separately</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>da_diff<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span> </span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> nFeature_RNA, <span class="at">y =</span> percent.mt, <span class="at">color =</span> is.outlier)) <span class="sc">+</span> </span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.6</span>) <span class="sc">+</span> </span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(sampleColors)) <span class="sc">+</span> </span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span>orig.ident) <span class="sc">+</span> </span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="scRNAseq-course_files/figure-html/quality-control2-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(da_diff, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"percent.mt"</span>),</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>        <span class="at">split.by =</span> <span class="st">"is.outlier"</span>,</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">ncol =</span> <span class="dv">3</span>,<span class="at">pt.size =</span> <span class="dv">0</span>) <span class="sc">&amp;</span> <span class="fu">scale_fill_manual</span>(<span class="at">values=</span>sampleColors)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="scRNAseq-course_files/figure-html/quality-control2-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="doublet-removal" class="level3">
<h3 class="anchored" data-anchor-id="doublet-removal">3.3 Doublet Removal</h3>
<p>Doublets occur when two cells are captured together during single-cell RNA sequencing, leading to artificial readouts that can distort the analysis. Detecting and removing doublets is essential to ensure the integrity of downstream analyses, as they can introduce noise and misleading biological signals.</p>
<p>In tissues with distinct cell types, such as the brain, cells can be categorized into well-defined populations based on specific gene expression profiles. Each cell type has unique markers that facilitate the identification of doublets, as the combination of two different cell types will typically produce a distinct expression profile that can be detected by computational methods. For example, if a doublet is formed between a neuron and a glial cell, the resulting expression profile may reveal markers from both cell types, allowing for effective identification.</p>
<p>In contrast, datasets with continuous maturation feature cells that may share similar gene expression profiles but exist at different developmental stages or functional states. For instance, in a dataset capturing the maturation of immune cells, the transitional states between precursor cells and fully mature cells can be subtle, making it difficult to distinguish between single cells and doublets based on gene expression alone. The absence of distinct markers complicates the identification process, as the expected combined expression patterns of doublets may closely resemble those of mature cells.</p>
<p>The table below compares the results of genetic demultiplexing (considered the ground truth) with the predictions of two doublet detection algorithms, <strong>DoubletFinder</strong> and <strong>scDblFinder</strong>, for two 10X samples derived from iDanoids (converted neurons, a mix of three cell lines). Rows represent the genetic demultiplexing results, while columns show the classification outcomes from the algorithms.</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th>Both</th>
<th>DoubletFinder</th>
<th>scDblFinder</th>
<th>Neither</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Doublet</td>
<td>76</td>
<td>1</td>
<td>71</td>
<td>60</td>
</tr>
<tr class="even">
<td>Singlet</td>
<td>85</td>
<td>246</td>
<td>260</td>
<td>7305</td>
</tr>
</tbody>
</table>
<p><strong>Question</strong>: Based on these results, do you find the predictions from these algorithms trustworthy?</p>
<section id="identifying-and-removing-doublets" class="level4">
<h4 class="anchored" data-anchor-id="identifying-and-removing-doublets">Identifying and Removing Doublets</h4>
<p>Nonetheless, we will proceed with running the analysis to familiarize ourselves with the method using the <strong>DoubletFinder</strong> package in R. Below is a script demonstrating how to use this method on your Seurat object to identify and filter out doublets from your dataset. For this analysis with need to normalize and scale the data and also perform dimensionality reduction. We will cover these in more detail later.</p>
<p>doubletFinder will add new columns to the meta data. We rename these for clarity. Finally we plot the predicted doublet status versus number of detected genes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalize data</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>da_diff <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(da_diff)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>da_diff <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(da_diff)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>da_diff <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(da_diff)</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>da_diff <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(da_diff, <span class="at">verbose=</span><span class="cn">FALSE</span>)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform doublet detection</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>da_diff <span class="ot">&lt;-</span> <span class="fu">doubletFinder</span>(da_diff, </span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">PCs =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>,</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">pK =</span> <span class="fl">0.1</span>, </span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">nExp =</span> <span class="fu">round</span>(<span class="fl">0.05</span> <span class="sc">*</span> <span class="fu">ncol</span>(da_diff)), </span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">reuse.pANN =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Creating 3385 artificial doublets..."
[1] "Creating Seurat object..."
[1] "Normalizing Seurat object..."
[1] "Finding variable genes..."
[1] "Scaling data..."
[1] "Running PCA..."
[1] "Calculating PC distance matrix..."
[1] "Computing pANN..."
[1] "Classifying doublets.."</code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(da_diff<span class="sc">@</span>meta.data) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(da_diff<span class="sc">@</span>meta.data) <span class="sc">%&gt;%</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  stringr<span class="sc">::</span><span class="fu">str_replace</span>(<span class="st">"pANN.*"</span>, <span class="st">"predicted_doublet_probability"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  stringr<span class="sc">::</span><span class="fu">str_replace</span>(<span class="st">"DF.classifications.*"</span>, <span class="st">"doublet_classification"</span>)</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>da_diff<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span> <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>nFeature_RNA, <span class="at">y=</span>nCount_RNA, <span class="at">color=</span>doublet_classification)) <span class="sc">+</span> <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">scale_color_manual</span>(<span class="at">values=</span><span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"red"</span>))<span class="sc">+</span><span class="fu">facet_wrap</span>(<span class="sc">~</span>orig.ident<span class="sc">+</span>doublet_classification)<span class="sc">+</span><span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="scRNAseq-course_files/figure-html/unnamed-chunk-1-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><strong>Question</strong>: One more harsh way of removing doublets is to exclude cells with number of genes detected &gt; than a manually assigned threshold. What are pros and cons of that approach?</p>
</section>
</section>
<section id="final-outlier-filtering" class="level3">
<h3 class="anchored" data-anchor-id="final-outlier-filtering">3.4 Final Outlier Filtering</h3>
<p>After identifying potential outliers and doublets, we can apply a final filtering step to exclude cells that are flagged as low quality or ambiguous. This final filter removes cells based on the previously calculated outlier status and doublet classification, ensuring that only high-quality, single-cell data remains for downstream analyses.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Final filtering step to remove outliers and doublets</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter out cells that were flagged as outliers or doublets</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>da_diff.filter<span class="ot">&lt;-</span> <span class="fu">subset</span>(da_diff, <span class="at">subset =</span> <span class="sc">!</span>outlier <span class="sc">&amp;</span> doublet_classification <span class="sc">!=</span> <span class="st">"Doublet"</span>)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Summary of cells retained after filtering</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Number of cells before final filtering:"</span>, <span class="fu">ncol</span>(da_diff)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Number of cells before final filtering: 10154"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Number of cells retained after final filtering:"</span>, <span class="fu">ncol</span>(da_diff.filter)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Number of cells retained after final filtering: 8793"</code></pre>
</div>
</div>
<p><strong>Question</strong>: What are the risks of having to harsh or to lenient filtering for removal of bad cells?</p>
<hr>
</section>
</section>
<section id="normalization-and-scaling" class="level1">
<h1>4. Normalization and Scaling</h1>
<section id="normalization" class="level3">
<h3 class="anchored" data-anchor-id="normalization">4.1 Normalization</h3>
<p>Normalization, finding variable features, and scaling are essential preprocessing steps to prepare the data for downstream analysis. Normalization corrects for differences in sequencing depth between cells, typically by transforming the raw counts into counts per cell or counts per million (CPM), followed by log transformation. This step ensures that expression levels are comparable across cells.</p>
<p>The total number of UMIs observed for a cell are attributable to</p>
<ol type="1">
<li><p>cell size (larger cell size =&gt; more RNA content =&gt; more expected UMIs) (less relevant for snRNA-seq).</p></li>
<li><p>Sequencing depth, which will have a large random component and can be modelled using a negative binomial distribution.</p></li>
</ol>
<p>Let’s look at the expression of a ubiquoulsy expressed gene (GAPDH):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>total_umis <span class="ot">&lt;-</span> da_diff.filter<span class="sc">$</span>nCount_RNA</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>gene <span class="ot">&lt;-</span> <span class="st">"GAPDH"</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>gene_umi <span class="ot">&lt;-</span> <span class="fu">GetAssayData</span>(da_diff.filter,<span class="at">layer =</span> <span class="st">"counts"</span>)[gene,]</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>expressionMatrix <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="st">"total_cell_umi"</span> <span class="ot">=</span> total_umis, <span class="st">"gene_umi"</span> <span class="ot">=</span> gene_umi)</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> expressionMatrix, </span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(total_cell_umi, gene_umi)) <span class="sc">+</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha=</span>.<span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Total sequencing depth"</span>) <span class="sc">+</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="fu">paste0</span>(<span class="st">"Gene UMI("</span>, gene,<span class="st">")"</span>))<span class="sc">+</span><span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>)<span class="sc">+</span><span class="fu">theme_bw</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="scRNAseq-course_files/figure-html/gapdh-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The gene exhibits an almost linear relationship with a cell’s sequencing depth. In this case, the “heterogeneity across cells” in GAPDH expression levels is primarily due to differences in sequencing depth rather than biological variability, making it less biologically interesting.</p>
<p>If all genes exhibited this linear behavior, we could simply account for sequencing depth by performing a linear regression of observed gene counts against each cell’s total sequencing depth. However, this approach is problematic because it would negatively impact the expression of marker genes, such as <em>NR4A2</em>, <em>COL1A1</em>, and <em>AQP4</em>. These marker genes are only expressed in specific subsets of cells, and applying linear regression to remove the effect of sequencing depth would risk dampening their expression, leading to a loss of biologically relevant information.</p>
<p>Instead of regressing out sequencing depth, an alternative approach is to use a global scaling factor (e.g., 10,000) to normalize all cells to a uniform sequencing depth while accounting for the original sequencing depth of each cell. After scaling the data, it’s common to apply a log-transformation. The log transformation helps compress the range of expression values (reducing the influence of highly expressed genes) and stabilizes the variance across genes. However, since many genes in scRNA-seq datasets have low or zero expression, you cannot apply the logarithm directly to zero values (since log⁡(0)(0)log(0) is undefined). To address this, a pseudocount (a small constant, typically 1) is added to all gene expression values before taking the logarithm. This ensures that all values are positive and that the log-transformation can be safely applied. In Seurat, this normalization can be performed using the <code>NormalizeData</code> function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalize data</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>da_diff.filter <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(da_diff.filter)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Normalizing layer: counts</code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>gene_umi_normalized <span class="ot">&lt;-</span> <span class="fu">GetAssayData</span>(da_diff.filter,<span class="at">layer =</span> <span class="st">"data"</span>)[gene,]</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> <span class="fu">data.frame</span>(<span class="st">"total_cell_umi"</span> <span class="ot">=</span> total_umis,</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>                         <span class="st">"gene_umi_norm"</span> <span class="ot">=</span> gene_umi_normalized), </span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(total_cell_umi, gene_umi_norm)) <span class="sc">+</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">14</span>) <span class="sc">+</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Total sequencing depth"</span>) <span class="sc">+</span></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="fu">paste0</span>(<span class="st">"Gene UMI("</span>, gene))<span class="sc">+</span><span class="fu">geom_smooth</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using method = 'gam' and formula = 'y ~ s(x, bs = "cs")'</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="scRNAseq-course_files/figure-html/normalization-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="finding-variable-features-scaling" class="level3">
<h3 class="anchored" data-anchor-id="finding-variable-features-scaling">4.2 Finding variable features &amp; Scaling</h3>
<p>Finding variable features identifies genes with significant expression variability across cells, which is important because these genes often capture biologically meaningful patterns, while others contribute more noise. Variable features are selected based on metrics like variance or mean-variance relationships, providing a focused subset of genes for analysis. Scaling centers and scales the expression data, often by subtracting the mean and dividing by the standard deviation of each gene across cells. This step standardizes the gene expression values, making it easier to compare cell-to-cell variation and enhancing the performance of algorithms used in downstream analyses like dimensionality reduction and clustering. Together, these steps improve the accuracy and interpretability of scRNA-seq data.</p>
<p>The <strong>ScaleData</strong> function is used to scale and center expression data. This process standardizes the data by shifting the distribution of each gene’s expression across cells to have a mean of zero and a standard deviation of one. ScaleData adjusts for cell-to-cell variability by centering and scaling each gene individually. Using ScaleData is particularly useful for analyses like PCA, which are sensitive to differences in gene variance. It also has the flexibility to regress out certain sources of variation, such as mitochondrial gene content, cell cycle phase, or other unwanted sources of noise.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify highly variable features</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>da_diff.filter <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(da_diff.filter, <span class="at">nfeatures =</span> <span class="dv">2000</span>)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="fu">VariableFeaturePlot</span>(da_diff.filter) <span class="sc">%&gt;%</span> </span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">LabelPoints</span>(<span class="at">points =</span> <span class="fu">head</span>(<span class="fu">VariableFeatures</span>(da_diff.filter),<span class="at">n=</span><span class="dv">20</span>), <span class="at">repel =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="scRNAseq-course_files/figure-html/variable_f-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Scaling data</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>da_diff.filter <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(da_diff.filter)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
</section>
<section id="dimensionality-reduction-and-clustering" class="level1">
<h1>5. Dimensionality Reduction and Clustering</h1>
<section id="principal-component-analysis-pca" class="level3">
<h3 class="anchored" data-anchor-id="principal-component-analysis-pca">5.1 Principal Component Analysis (PCA)</h3>
<p>After quality control, filtering and normalization, the next step in single-cell RNA sequencing analysis is usually to explore the structure of the data by grouping similar cells and visualizing these groupings in a reduced-dimensional space. This is achieved through <strong>dimensionality reduction</strong> and <strong>clustering</strong>.</p>
<p>Dimensionality reduction techniques help to condense complex, high-dimensional scRNA-seq data into fewer dimensions, making it easier to identify patterns and relationships among cells. Clustering algorithms then group cells with similar gene expression profiles, revealing potential cell types or states present in the dataset.</p>
<p><strong>Principal Component Analysis (PCA)</strong> is one of the most commonly used dimensionality reduction methods in scRNA-seq analysis. PCA transforms the data by identifying principal components, which capture the axes of maximum variance in the data, thereby highlighting major sources of variation. Here, we run PCA on the filtered dataset using the most variable genes to capture the essential features of the data, selecting 50 principal components (PCs) to examine. The ElbowPlot function helps determine the number of PCs to retain by visualizing the point where adding more components provides diminishing returns in variance explained.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run PCA</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>da_diff.filter <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(da_diff.filter, </span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>                         <span class="at">features =</span> <span class="fu">VariableFeatures</span>(<span class="at">object =</span> da_diff.filter),</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>                        <span class="at">npcs =</span> <span class="dv">50</span>,</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>                        <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize PCA</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ElbowPlot</span>(da_diff.filter, <span class="at">ndims =</span> <span class="dv">50</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="scRNAseq-course_files/figure-html/pca-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="clustering-and-visualization" class="level3">
<h3 class="anchored" data-anchor-id="clustering-and-visualization">5.2 Clustering and Visualization</h3>
<p>Once we have a reduced representation of the data from PCA, we proceed with clustering to group cells based on their similarity. In this workflow, we use the <code>FindNeighbors</code> function to calculate the local neighborhood structure of cells based on their PCA coordinates, followed by <code>FindClusters</code> to assign each cell to a cluster. The resolution parameter is set to 0.5 to control the granularity of clusters, with higher values yielding more clusters.</p>
<p>In constructing the local neighborhood structure, <code>FindNeighbors</code> employs a <strong>Shared Nearest Neighbor (SNN)</strong> approach to refine the clustering accuracy. The SNN method goes beyond simple k-nearest neighbors by assessing the overlap of each cell’s neighborhood with those of other cells. Cells are considered more similar if they share multiple nearest neighbors, resulting in a <strong>weighted graph</strong> that highlights stronger connections between cells with high shared neighbor counts. This added layer of similarity enhances clustering stability by prioritizing dense regions of cells that share local neighborhoods, thereby emphasizing biologically relevant groupings in complex data.</p>
<p>To visualize these clusters, we apply Uniform Manifold Approximation and Projection (UMAP), which provides a 2D projection of the data, preserving local and some global structure in the clustering. The UMAP plot provides an intuitive visual of how cells are grouped, making it easier to explore potential cell types or states within the dataset. However, the UMAP should not be overinterpreted. UMAP performs dimensionality reduction, which inherently simplifies the data. While this reduction is helpful for visualizing high-dimensional scRNA-seq datasets, it inevitably loses some information. Rare cell populations or subtle differences between similar cell states may not be adequately captured in the UMAP embedding. Unlike principal component analysis (PCA), where axes represent specific directions of variance, UMAP’s axes are not directly tied to biologically meaningful dimensions. As a result, the spatial arrangement of clusters should be interpreted with caution. For instance, the distance between clusters in UMAP space does not always reflect the true degree of similarity or dissimilarity between cell populations. Lastly, UMAP results are influenced by its parameters, such as the number of neighbors and the minimum distance. These parameters affect the resolution of the embedding and can alter the apparent number and shape of clusters.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Find clusters</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>da_diff.filter <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(da_diff.filter, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>)</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>da_diff.filter <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(da_diff.filter, <span class="at">resolution =</span> <span class="fl">0.1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 8793
Number of edges: 307533

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.9598
Number of communities: 8
Elapsed time: 1 seconds</code></pre>
</div>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run UMAP for visualization</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>da_diff.filter <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(da_diff.filter, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(da_diff.filter, <span class="at">reduction =</span> <span class="st">"umap"</span>, <span class="at">group.by =</span> <span class="st">"seurat_clusters"</span>,<span class="at">cols=</span>clusterColors)<span class="sc">+</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">DimPlot</span>(da_diff.filter, <span class="at">group.by =</span> <span class="st">"orig.ident"</span>, <span class="at">reduction =</span> <span class="st">"umap"</span>,<span class="at">cols =</span> sampleColors)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="scRNAseq-course_files/figure-html/clustering-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="integration-using-harmony" class="level3">
<h3 class="anchored" data-anchor-id="integration-using-harmony">5.3 Integration Using Harmony</h3>
<p>Batch correction in scRNA-seq is necessary when technical variation influences the data, masking biological differences. This is critical if samples from different conditions or groups were processed in separate batches, as variations in sequencing runs or reagents can introduce unwanted noise. Knowing when to apply batch correction depends on the experimental design, data structure, and the specific downstream analyses planned. Importantly, batch correction should never replace proper experimtal design and planning. Running all “treated” samples on day 1 and then on day 2 run all “control” samples is an example of poor experimental planing.</p>
<p>The harmony algorithm provides an effective approach to integrating data from different sources while controlling for these batch effects. Harmony works by iteratively adjusting the embeddings of cells in a lower-dimensional space, aligning them based on shared biological variation while minimizing differences caused by batch effects. This is achieved through a low-rank approximation of the data, which allows for the simultaneous consideration of both biological and technical sources of variation.</p>
<p>To integrate datasets using Harmony, we start by running PCA on each dataset separately. Once we have the PCA embeddings, we can pass them to the RunHarmony function, specifying the variable that represents the batch or dataset. Harmony then re-embeds the cells, producing a new PCA representation that mitigates batch effects while retaining the biological structure.</p>
<p>After integrating the datasets with Harmony, we can proceed with clustering and visualization using the adjusted embeddings. This integration enhances the robustness of downstream analyses, allowing for more accurate identification of cell types or states across datasets, and ensuring that biological signals are not overshadowed by batch variability.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Integrate using Harmony</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>da_integrated <span class="ot">&lt;-</span> <span class="fu">RunHarmony</span>(da_diff.filter, <span class="at">group.by.vars =</span> <span class="st">"orig.ident"</span>)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Proceed with clustering and visualization on the Harmony-adjusted embeddings</span></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>da_integrated <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(da_integrated, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>, <span class="at">reduction =</span> <span class="st">"harmony"</span>)</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>da_integrated <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(da_integrated, <span class="at">resolution =</span> <span class="fl">0.1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck

Number of nodes: 8793
Number of edges: 354160

Running Louvain algorithm...
Maximum modularity in 10 random starts: 0.9598
Number of communities: 7
Elapsed time: 1 seconds</code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>da_integrated <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(da_integrated, <span class="at">reduction =</span> <span class="st">"harmony"</span>,</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>                         <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>,<span class="at">reduction.name =</span> <span class="st">"umap_harmony"</span>)</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize the integrated data</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(da_integrated, <span class="at">reduction =</span> <span class="st">"umap_harmony"</span>, <span class="at">group.by =</span> <span class="st">"seurat_clusters"</span>, <span class="at">cols=</span>clusterColors) <span class="sc">+</span></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(da_integrated, <span class="at">group.by =</span> <span class="st">"orig.ident"</span>, <span class="at">reduction =</span> <span class="st">"umap_harmony"</span>, <span class="at">cols =</span> sampleColors)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="scRNAseq-course_files/figure-html/harmony-integration-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<hr>
</section>
</section>
<section id="saving-your-processed-object" class="level1">
<h1>6. Saving your processed object</h1>
<p>At this stage, it’s a good idea to save your Seurat object. This allows you to pick up your analysis directly from this point in the future, whether revisiting your project or sharing it with a colleague. To do this, we’ll save the object as an RDS file. The RDS format is a standard way to store R objects, preserving all the data and metadata associated with your analysis.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">"processed_data/"</span>)</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(da_integrated, <span class="st">"processed_data/da_integrated.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To reload the object in the future, you can use the following command:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load the saved Seurat object</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>da_integrated <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"processed_data/da_integrated.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="annotating-clusters" class="level1">
<h1>7. Annotating Clusters</h1>
<p>Clustering cells assigns an identity label to each one, allowing us to infer that cells with the same label are similar and likely represent the same cell type or state. Determining the specific cell types or states corresponding to these clusters can be challenging, and there is often no definitive answer. However, several approaches can help clarify this issue:</p>
<ol type="1">
<li><p><strong>Examine the expression of canonical cell type and cell state markers</strong> within the clusters. This method relies on established knowledge of well-characterized markers in the field.</p></li>
<li><p><strong>Identify signature or marker genes</strong> for each cell cluster (i.e. running differential expression analysis between clusters). Once these genes are determined, literature searches, enrichment analyses, or experimental validations can be conducted to further refine annotations.</p></li>
<li><p><strong>Compare the gene expression profiles of the clusters</strong> with existing reference datasets using automated methods. This can provide additional context for identifying the nature of the clusters.</p></li>
</ol>
<p>It is important to note that the first method requires prior knowledge of the system being studied, including a comprehensive list of accepted markers relevant to the context. For instance, in the context of differentiating dopamine neurons from stem cells, several suitable marker genes have been identified, including:</p>
<ul>
<li>MKI67, TOP2A, CENPF: Genes involved in the cell cycle, often used to mark proliferating cells.</li>
<li>SOX2, HES1, NES: Key genes associated with neurogenesis, marking neural stem or progenitor cells.</li>
<li>FABP7, SHH, CORIN, FOXA2, LMX1B, OTX2: Markers of floor plate progenitors, important for specifying ventral midbrain neurons.</li>
<li>TH, EN1, PBX1, STMN2: Genes involved in dopamine (DA) neurogenesis; critical for identifying dopamine neuron lineage and development.</li>
<li>AQP4, GFAP, EDNRB, GJA1: Markers of astrocytes, essential for identifying and studying astrocytic populations.</li>
<li>COL1A1, COL1A2, LUM: Genes associated with vascular and leptomeningeal cells (VLMCs), playing a role in extracellular matrix formation and support.</li>
</ul>
<p>These genes serve as a starting point for assessing the identity of midbrain progenitor derivatives in clusters derived from stem cell differentiation.</p>
<section id="canonical-markers" class="level4">
<h4 class="anchored" data-anchor-id="canonical-markers">7.1 Canonical Markers</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Canonical markers for dopamine neurons</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">"MKI67"</span>,<span class="st">"TOP2A"</span>,<span class="st">"CENPF"</span>, <span class="co"># Cell cycle</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">"SOX2"</span>, <span class="st">"HES1"</span>, <span class="st">"NES"</span>,         <span class="co"># Neurogenesis</span></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>  <span class="st">"FABP7"</span>, <span class="st">"SHH"</span>, <span class="st">"CORIN"</span>, <span class="st">"FOXA2"</span>, <span class="st">"LMX1B"</span>, <span class="st">"OTX2"</span>, <span class="co"># Floor plate progenitors</span></span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>  <span class="st">"TH"</span>, <span class="st">"EN1"</span>, <span class="st">"PBX1"</span>, <span class="st">"STMN2"</span>,           <span class="co"># DA neurogenesis</span></span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>  <span class="st">"AQP4"</span>, <span class="st">"GFAP"</span>, <span class="st">"EDNRB"</span>,        <span class="co"># Astrocyte</span></span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>  <span class="st">"COL1A1"</span>, <span class="st">"COL1A2"</span>, <span class="st">"LUM"</span>   <span class="co"># VLMC</span></span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Violin plot for canonical markers</span></span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a><span class="fu">DotPlot</span>(da_integrated, <span class="at">features =</span> genes,<span class="at">dot.scale =</span> <span class="dv">8</span>)<span class="sc">+</span></span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a>  <span class="co">#Rotate x axis labels</span></span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">hjust =</span> <span class="dv">1</span>))<span class="sc">+</span></span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Update colors to something prettier</span></span>
<span id="cb34-16"><a href="#cb34-16" aria-hidden="true" tabindex="-1"></a>viridis<span class="sc">::</span><span class="fu">scale_color_viridis</span>(<span class="at">direction =</span> <span class="sc">-</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Scale for colour is already present.
Adding another scale for colour, which will replace the existing scale.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="scRNAseq-course_files/figure-html/canonical-markers-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Instead of looking at one marker at a time we can also create list(s) of relevant genes and calculate a module score for each cluster. This can help us identify the most relevant genes for each cluster and potentially identify new markers.</p>
<p>In this example we are using Gene Ontology terms to create gene modules for neurogenesis, cell cycle and stress response. We then calculate a module score for each cluster and visualize the results.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Download gene modules from GO related to neurogenesis, cell cycle and stress</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>gene_modules <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Generation of cells within the nervous system</span></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">Neurogenesis=</span><span class="fu">bitr</span>(<span class="st">"GO:0071542"</span>, <span class="at">fromType=</span><span class="st">"GOALL"</span>, <span class="at">toType=</span><span class="st">"SYMBOL"</span>, <span class="at">OrgDb=</span><span class="st">'org.Hs.eg.db'</span>) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(<span class="st">"SYMBOL"</span>), </span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Cellular response to hypoxia</span></span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">Stress_Response=</span><span class="fu">bitr</span>(<span class="st">"GO:0071456"</span>, <span class="at">fromType=</span><span class="st">"GOALL"</span>, <span class="at">toType=</span><span class="st">"SYMBOL"</span>, <span class="at">OrgDb=</span><span class="st">'org.Hs.eg.db'</span>) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(<span class="st">"SYMBOL"</span>),</span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Cell cycle</span></span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">Cell_Cycle=</span><span class="fu">bitr</span>(<span class="st">"GO:0007049"</span>, <span class="at">fromType=</span><span class="st">"GOALL"</span>, <span class="at">toType=</span><span class="st">"SYMBOL"</span>, <span class="at">OrgDb=</span><span class="st">'org.Hs.eg.db'</span>) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(<span class="st">"SYMBOL"</span>))</span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="fu">names</span>(gene_modules)) {</span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a>  da_integrated <span class="ot">&lt;-</span> <span class="fu">AddModuleScore</span>(<span class="at">object =</span> da_integrated, </span>
<span id="cb36-13"><a href="#cb36-13" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">features =</span> <span class="fu">list</span>(gene_modules[[i]]), </span>
<span id="cb36-14"><a href="#cb36-14" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">name =</span> i)</span>
<span id="cb36-15"><a href="#cb36-15" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb36-16"><a href="#cb36-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-17"><a href="#cb36-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Add module score will add the score in the metadata with the name of the list but append 1 to the name.</span></span>
<span id="cb36-18"><a href="#cb36-18" aria-hidden="true" tabindex="-1"></a><span class="co"># As we have seen before the VlnPlot function can be used to visualise features in the metadata</span></span>
<span id="cb36-19"><a href="#cb36-19" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(</span>
<span id="cb36-20"><a href="#cb36-20" aria-hidden="true" tabindex="-1"></a>    da_integrated,</span>
<span id="cb36-21"><a href="#cb36-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">features =</span> <span class="fu">paste0</span>(<span class="fu">names</span>(gene_modules),<span class="st">"1"</span>),</span>
<span id="cb36-22"><a href="#cb36-22" aria-hidden="true" tabindex="-1"></a>    <span class="at">pt.size =</span> <span class="dv">0</span>, <span class="at">cols =</span> clusterColors)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="scRNAseq-course_files/figure-html/module-markers-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="identifying-marker-genes" class="level4">
<h4 class="anchored" data-anchor-id="identifying-marker-genes">7.2 Identifying Marker Genes</h4>
<p>Marker genes are genes that are differentially expressed in specific cell populations or clusters and can be used to characterize the biological functions and identities of these populations. In scRNA-seq analysis, marker genes are typically identified by comparing the expression levels between clusters of cells, often using statistical tests to find genes that are uniquely expressed in particular groups.</p>
<p>Historically, the most common used method has been yhe Wilcoxon test on single cells which treats each cell as an independent observation, comparing gene expression distributions between clusters. This method is sensitive to subtle differences in cell states and does not require aggregating data, making it straightforward to implement. However, it assumes cells are independent, which can overlook the biological reality that cells within the same cluster or sample are correlated. This can lead to inflated statistical significance due to the large number of cells.</p>
<p>More recentyl, pseudobulk approaches has gained traction. Pseudobulk methods aggregate the expression data within each cluster or sample before performing differential expression analysis. By treating each sample as a unit, pseudobulk analysis accounts for biological replication and reduces the impact of technical noise, usually providing more robust and reproducible results.</p>
<p>Here we will use the Wilcoxon test as implemented in the Seurat funciton <em>FindAllMarkers</em> to identify marker genes for each cluster. We will then visualize the top 10 marker genes for each cluster using a heatmap.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Find markers for each cluster</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>da_diff.markers <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(da_integrated, <span class="at">only.pos =</span> <span class="cn">TRUE</span>,</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">min.pct =</span> <span class="fl">0.25</span>, <span class="at">logfc.threshold =</span> <span class="fl">0.25</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 0</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 1</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 4</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 5</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Calculating cluster 6</code></pre>
</div>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># downsample to 100 cells per cluster</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>da_integrated.subsampled.metadata <span class="ot">&lt;-</span> da_integrated<span class="sc">@</span>meta.data <span class="sc">%&gt;%</span> <span class="fu">rownames_to_column</span>() <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(seurat_clusters) <span class="sc">%&gt;%</span> <span class="fu">sample_n</span>(<span class="dv">50</span>) </span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a><span class="co"># top 20 genes per cluster</span></span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>da_diff.markers.top <span class="ot">&lt;-</span> da_diff.markers <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(cluster) <span class="sc">%&gt;%</span> <span class="fu">top_n</span>(<span class="dv">5</span>, avg_log2FC) <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(cluster, avg_log2FC) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(gene)</span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Heatmap for canonical markers</span></span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>da_integrated <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(da_integrated, <span class="at">features =</span> da_diff.markers.top)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Centering and scaling data matrix</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Different features in new layer data than already exists for
scale.data</code></pre>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DoHeatmap</span>(da_integrated, <span class="at">features =</span> da_diff.markers.top, <span class="at">group.by =</span> <span class="st">"seurat_clusters"</span>,<span class="at">cells =</span>da_integrated.subsampled.metadata<span class="sc">$</span>rowname)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="scRNAseq-course_files/figure-html/markers-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="automated-annotation-using-singler" class="level4">
<h4 class="anchored" data-anchor-id="automated-annotation-using-singler">7.3.1 Automated annotation using singleR</h4>
<p>The third way that we will try out to classify cells is by using reference-based classification methods, which rely on prior knowledge of gene expression profiles for known cell types. SingleR is a powerful tool for this task, which performs reference-based classification of single-cell data by comparing the gene expression profiles of the cells in your dataset to a reference dataset of annotated cell types.</p>
<p>In this analysis, we use the LaMannoBrainData reference dataset, which contains gene expression data for human embryonic brain cell types, to classify cells from a different dataset, the data comes from <a href="https://pubmed.ncbi.nlm.nih.gov/27716510/">this paper</a>. The classification results are then added to the Seurat object as metadata, allowing us to visualize and analyze the cell type predictions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Prepare la manno data set</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>la_manno_ref <span class="ot">&lt;-</span> <span class="fu">LaMannoBrainData</span>(<span class="at">which =</span> <span class="st">"human-embryo"</span>)</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a><span class="fu">logcounts</span>(la_manno_ref) <span class="ot">&lt;-</span> <span class="fu">log2</span>(<span class="fu">counts</span>(la_manno_ref)<span class="sc">+</span><span class="dv">1</span>)</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Run SingleR to classify cells</span></span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>singleR_results <span class="ot">&lt;-</span> <span class="fu">SingleR</span>(</span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">test =</span> <span class="fu">as.SingleCellExperiment</span>(da_integrated),</span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">ref =</span> la_manno_ref,  </span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">labels =</span> la_manno_ref<span class="sc">$</span>Cell_type</span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb49-11"><a href="#cb49-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-12"><a href="#cb49-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Add SingleR predictions as metadata to Seurat object</span></span>
<span id="cb49-13"><a href="#cb49-13" aria-hidden="true" tabindex="-1"></a>da_integrated<span class="sc">$</span>SingleR_label <span class="ot">&lt;-</span> singleR_results<span class="sc">$</span>labels</span>
<span id="cb49-14"><a href="#cb49-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-15"><a href="#cb49-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize SingleR predictions</span></span>
<span id="cb49-16"><a href="#cb49-16" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(da_integrated, <span class="at">group.by =</span> <span class="st">"SingleR_label"</span>, <span class="at">reduction =</span> <span class="st">"umap_harmony"</span>,<span class="at">label =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="scRNAseq-course_files/figure-html/singler-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>To further refine the cell type classification, we can filter out cell types with low representation in the dataset. This step helps focus the analysis on the most prevalent and well-characterized cell types, improving the interpretability of the results.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate percentage of each cell type</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>cell_type_counts <span class="ot">&lt;-</span> <span class="fu">table</span>(da_integrated<span class="sc">$</span>SingleR_label)</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Update metadata to include only cell types above threshold</span></span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a>da_integrated<span class="sc">$</span>Filtered_SingleR_label <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(</span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a>  da_integrated<span class="sc">$</span>SingleR_label <span class="sc">%in%</span> <span class="fu">names</span>(cell_type_counts[cell_type_counts <span class="sc">&gt;</span> <span class="dv">200</span>]),</span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a>  da_integrated<span class="sc">$</span>SingleR_label,</span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a>  <span class="cn">NA</span>  <span class="co"># or use "Other"</span></span>
<span id="cb50-9"><a href="#cb50-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb50-10"><a href="#cb50-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-11"><a href="#cb50-11" aria-hidden="true" tabindex="-1"></a><span class="co"># UMAP plot with filtered cell types</span></span>
<span id="cb50-12"><a href="#cb50-12" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(da_integrated, <span class="at">group.by =</span> <span class="st">"Filtered_SingleR_label"</span>,<span class="at">reduction =</span> <span class="st">"umap_harmony"</span>,  <span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">na.value =</span> <span class="st">"grey"</span>) <span class="sc">+</span> </span>
<span id="cb50-13"><a href="#cb50-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggtitle</span>(<span class="st">"Filtered Cell Type Classification ( &gt; 200 cells)"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="scRNAseq-course_files/figure-html/singler_filter-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="automated-annotation-using-seurats-labeltransfer" class="level4">
<h4 class="anchored" data-anchor-id="automated-annotation-using-seurats-labeltransfer">7.3.2 Automated annotation using Seurats labeltransfer</h4>
<p>Seurat’s Label Transfer function is another method used to transfer cell type annotations or other labels from a reference dataset onto a query dataset in single-cell RNA-seq analysis. Label Transfer works by first aligning the datasets through the integration of shared biological features, then identifying anchor points (common structures or cell states) between the reference and query. Using these anchors, Seurat calculates a prediction score for each label in the reference dataset and transfers the highest-scoring labels to the query cells. This method maintains biological consistency between datasets while accounting for batch effects or differences in sequencing protocols, making it a powerful tool for harmonizing annotations across single-cell studies.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>la_manno.seu <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> <span class="fu">counts</span>(la_manno_ref), <span class="at">project =</span> <span class="st">"LaMannoBrainData"</span>,</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">meta.data =</span> <span class="fu">data.frame</span>(<span class="fu">colData</span>(la_manno_ref)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Feature names cannot have underscores ('_'), replacing with dashes
('-')</code></pre>
</div>
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>la_manno.seu <span class="ot">&lt;-</span> la_manno.seu[,la_manno.seu<span class="sc">$</span>Cell_type<span class="sc">!=</span><span class="st">"Unk"</span>]</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>la_manno.seu <span class="ot">&lt;-</span> la_manno.seu <span class="sc">%&gt;%</span> <span class="fu">NormalizeData</span>() <span class="sc">%&gt;%</span> <span class="fu">FindVariableFeatures</span>(<span class="at">nfeatures =</span> <span class="dv">4000</span>) <span class="sc">%&gt;%</span> <span class="fu">ScaleData</span>() <span class="sc">%&gt;%</span> <span class="fu">RunPCA</span>(<span class="at">npcs =</span> <span class="dv">30</span>, <span class="at">verbose =</span> F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Normalizing layer: counts</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding variable features for layer counts</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Centering and scaling data matrix</code></pre>
</div>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Find anchors between the reference and query datasets</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>anchors <span class="ot">&lt;-</span> <span class="fu">FindTransferAnchors</span>(<span class="at">reference =</span> la_manno.seu, <span class="at">query =</span> da_integrated, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Performing PCA on the provided reference using 3148 features as input.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Projecting cell embeddings</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding neighborhoods</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding anchors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>    Found 1130 anchors</code></pre>
</div>
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Transfer labels from the reference to the query dataset</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>predictions <span class="ot">&lt;-</span> <span class="fu">TransferData</span>(<span class="at">anchorset =</span> anchors, <span class="at">refdata =</span>la_manno.seu<span class="sc">$</span>Cell_type)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding integration vectors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Finding integration vector weights</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Predicting cell labels</code></pre>
</div>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Add transferred labels to the query dataset</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>da_integrated <span class="ot">&lt;-</span> <span class="fu">AddMetaData</span>(da_integrated, <span class="at">metadata =</span> predictions)</span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Lets remove cells only found in low numbers</span></span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a>cell_type_counts_label_transfer <span class="ot">&lt;-</span> <span class="fu">table</span>(predictions<span class="sc">$</span>predicted.id)</span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-8"><a href="#cb67-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Update metadata to include only cell types above threshold</span></span>
<span id="cb67-9"><a href="#cb67-9" aria-hidden="true" tabindex="-1"></a>da_integrated<span class="sc">$</span>filtered_label_transfer <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(</span>
<span id="cb67-10"><a href="#cb67-10" aria-hidden="true" tabindex="-1"></a>  da_integrated<span class="sc">$</span>predicted.id <span class="sc">%in%</span> <span class="fu">names</span>(cell_type_counts_label_transfer[cell_type_counts_label_transfer <span class="sc">&gt;</span> <span class="dv">200</span>]),</span>
<span id="cb67-11"><a href="#cb67-11" aria-hidden="true" tabindex="-1"></a>  da_integrated<span class="sc">$</span>predicted.id,</span>
<span id="cb67-12"><a href="#cb67-12" aria-hidden="true" tabindex="-1"></a>  <span class="cn">NA</span>  <span class="co"># or use "Other"</span></span>
<span id="cb67-13"><a href="#cb67-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb67-14"><a href="#cb67-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-15"><a href="#cb67-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-16"><a href="#cb67-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize the transferred labels</span></span>
<span id="cb67-17"><a href="#cb67-17" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(da_integrated, <span class="at">group.by =</span> <span class="st">"filtered_label_transfer"</span>, <span class="at">reduction =</span> <span class="st">"umap_harmony"</span>, <span class="at">label =</span> T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="scRNAseq-course_files/figure-html/label_t-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The prediction.score.max in Seurat’s label transfer reflects the confidence of a predicted label for a query cell, ranging from 0 (low confidence) to 1 (high confidence). Higher scores indicate a strong match between the query cell and a reference cell type, while lower scores suggest poor matches, potential novel cell types, or technical noise. You can use thresholds to filter low-confidence cells or analyze them further for biological insights. Proper interpretation depends on the quality of the reference dataset and the context of your analysis.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(da_integrated, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"prediction.score.max"</span>),<span class="at">order=</span>T,<span class="at">min.cutoff =</span> <span class="st">"q9"</span>,<span class="at">reduction =</span> <span class="st">"umap_harmony"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="scRNAseq-course_files/figure-html/label_t2-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<hr>
</section>
</section>
<section id="extras" class="level1">
<h1>8. Extras</h1>
<section id="pathway-analysis-with-clusterprofiler" class="level3">
<h3 class="anchored" data-anchor-id="pathway-analysis-with-clusterprofiler">8.1 Pathway Analysis with clusterProfiler</h3>
<p>In this section, we perform pathway enrichment analysis using marker genes identified in <strong>Section 7.2</strong>. The <strong>clusterProfiler</strong> package allows us to investigate biological pathways associated with clusters or conditions of interest.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter significant markers for a specific cluster (e.g., Cluster 1)</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>cluster5_markers <span class="ot">&lt;-</span> da_diff.markers <span class="sc">%&gt;%</span> </span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(cluster <span class="sc">==</span> <span class="st">"5"</span> <span class="sc">&amp;</span> p_val_adj <span class="sc">&lt;</span> <span class="fl">0.05</span>) <span class="sc">%&gt;%</span>  <span class="co"># Adjust for your cluster and significance threshold</span></span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(gene)</span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert gene symbols to Entrez IDs</span></span>
<span id="cb69-7"><a href="#cb69-7" aria-hidden="true" tabindex="-1"></a>entrez_ids <span class="ot">&lt;-</span> <span class="fu">bitr</span>(cluster5_markers, <span class="at">fromType =</span> <span class="st">"SYMBOL"</span>, </span>
<span id="cb69-8"><a href="#cb69-8" aria-hidden="true" tabindex="-1"></a>                   <span class="at">toType =</span> <span class="st">"ENTREZID"</span>, </span>
<span id="cb69-9"><a href="#cb69-9" aria-hidden="true" tabindex="-1"></a>                   <span class="at">OrgDb =</span> org.Hs.eg.db)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in bitr(cluster5_markers, fromType = "SYMBOL", toType = "ENTREZID", :
3.61% of input gene IDs are fail to map...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="co"># GO enrichment analysis</span></span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>go_results <span class="ot">&lt;-</span> <span class="fu">enrichGO</span>(<span class="at">gene =</span> entrez_ids<span class="sc">$</span>ENTREZID,</span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">OrgDb =</span> org.Hs.eg.db, </span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a>                       <span class="at">keyType =</span> <span class="st">"ENTREZID"</span>, </span>
<span id="cb72-5"><a href="#cb72-5" aria-hidden="true" tabindex="-1"></a>                       <span class="at">ont =</span> <span class="st">"BP"</span>,         <span class="co"># Biological Process ontology</span></span>
<span id="cb72-6"><a href="#cb72-6" aria-hidden="true" tabindex="-1"></a>                       <span class="at">pAdjustMethod =</span> <span class="st">"BH"</span>, </span>
<span id="cb72-7"><a href="#cb72-7" aria-hidden="true" tabindex="-1"></a>                       <span class="at">pvalueCutoff =</span> <span class="fl">0.05</span>)</span>
<span id="cb72-8"><a href="#cb72-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-9"><a href="#cb72-9" aria-hidden="true" tabindex="-1"></a><span class="co"># View summary of enriched GO terms</span></span>
<span id="cb72-10"><a href="#cb72-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Dot plot</span></span>
<span id="cb72-11"><a href="#cb72-11" aria-hidden="true" tabindex="-1"></a><span class="fu">dotplot</span>(go_results, <span class="at">showCategory =</span> <span class="dv">10</span>) <span class="sc">+</span> </span>
<span id="cb72-12"><a href="#cb72-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Top 10 GO Terms for Biological Processes"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="scRNAseq-course_files/figure-html/pw1-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="co"># KEGG pathway enrichment analysis</span></span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>kegg_results <span class="ot">&lt;-</span> <span class="fu">enrichKEGG</span>(<span class="at">gene =</span> entrez_ids<span class="sc">$</span>ENTREZID,</span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>                           <span class="at">organism =</span> <span class="st">'hsa'</span>,       <span class="co"># Use 'mmu' for mouse</span></span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a>                           <span class="at">pvalueCutoff =</span> <span class="fl">0.05</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Reading KEGG annotation online: "https://rest.kegg.jp/link/hsa/pathway"...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Reading KEGG annotation online: "https://rest.kegg.jp/list/pathway/hsa"...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize KEGG results</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dotplot</span>(kegg_results, <span class="at">showCategory =</span> <span class="dv">10</span>) <span class="sc">+</span></span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Top 10 Enriched KEGG Pathways"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="scRNAseq-course_files/figure-html/pw1-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="identifying-number-of-clusters" class="level3">
<h3 class="anchored" data-anchor-id="identifying-number-of-clusters">8.2 Identifying number of clusters</h3>
<p>The resolution parameter in Seurat’s FindClusters function controls the granularity of clustering, with higher resolutions producing more fine-grained clusters. However, determining the optimal resolution can be challenging, especially when biological relevance is not immediately apparent.</p>
<p>The clustree package provides a powerful visualization tool to assess how clusters evolve across different resolutions. By creating a tree-like structure, clustree highlights how clusters merge or split as resolution changes, aiding in the identification of biologically meaningful patterns. This section demonstrates how to generate clustering solutions at multiple resolutions and visualize them using clustree.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load required libraries</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages(c("clustree"))</span></span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(clustree)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: ggraph</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'ggraph'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:sp':

    geometry</code></pre>
</div>
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform clustering for a range of resolutions</span></span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>resolutions <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="fl">0.1</span>, <span class="dv">1</span>, <span class="at">by =</span> <span class="fl">0.2</span>)         <span class="co"># Define a range of resolutions</span></span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a>da_diff <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(da_integrated, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="at">reduction =</span> <span class="st">"harmony"</span>) <span class="co"># Identify cell neighbors</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing nearest neighbor graph</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Computing SNN</code></pre>
</div>
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (res <span class="cf">in</span> resolutions) {</span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a>  da_integrated <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(</span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a>    da_integrated,</span>
<span id="cb84-4"><a href="#cb84-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">resolution =</span> res,                          <span class="co"># Specify the current resolution</span></span>
<span id="cb84-5"><a href="#cb84-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">verbose =</span> <span class="cn">FALSE</span>                            <span class="co"># Suppress output</span></span>
<span id="cb84-6"><a href="#cb84-6" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb84-7"><a href="#cb84-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb84-8"><a href="#cb84-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-9"><a href="#cb84-9" aria-hidden="true" tabindex="-1"></a><span class="co"># clustree visualization</span></span>
<span id="cb84-10"><a href="#cb84-10" aria-hidden="true" tabindex="-1"></a><span class="fu">clustree</span>(da_integrated<span class="sc">@</span>meta.data, <span class="at">prefix =</span> <span class="st">"RNA_snn_res."</span>) <span class="sc">+</span></span>
<span id="cb84-11"><a href="#cb84-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb84-12"><a href="#cb84-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Clustering Resolution Tree"</span>) <span class="sc">+</span></span>
<span id="cb84-13"><a href="#cb84-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Resolution"</span>) <span class="sc">+</span></span>
<span id="cb84-14"><a href="#cb84-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Clusters"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="scRNAseq-course_files/figure-html/clustree-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>In addition to clustree, other methods like silhouette analysis can help assess the quality and biological relevance of clustering results. Silhouette analysis provides a metric that quantifies how well a cell fits within its assigned cluster compared to neighboring clusters. A high silhouette width indicates well-separated clusters, while a low or negative silhouette width suggests overlapping or poorly defined clusters.</p>
<p>Silhouette anaylis can be used both to asses the stability of a given clustering results but also to find the optimal number of clusters by optimizing the overall silhouette score.</p>
<p>Here we will use silhouette analysis to look at the stability of the clustering results at a given resolution.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load required libraries</span></span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cluster)</span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract the cluster assignments and the PCA coordinates</span></span>
<span id="cb85-5"><a href="#cb85-5" aria-hidden="true" tabindex="-1"></a>clusters <span class="ot">&lt;-</span> da_integrated<span class="sc">$</span>RNA_snn_res.<span class="fl">0.1</span>              <span class="co"># Cluster labels, here with resolution 0.1</span></span>
<span id="cb85-6"><a href="#cb85-6" aria-hidden="true" tabindex="-1"></a>coords <span class="ot">&lt;-</span> <span class="fu">Embeddings</span>(da_integrated, <span class="at">reduction =</span> <span class="st">"harmony"</span>)[, <span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>] <span class="co"># PCA coordinates (first 10 dimensions)</span></span>
<span id="cb85-7"><a href="#cb85-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-8"><a href="#cb85-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute the silhouette width</span></span>
<span id="cb85-9"><a href="#cb85-9" aria-hidden="true" tabindex="-1"></a>sil <span class="ot">&lt;-</span> <span class="fu">silhouette</span>(<span class="fu">as.numeric</span>(clusters)<span class="sc">-</span><span class="dv">1</span>, <span class="fu">dist</span>(coords))</span>
<span id="cb85-10"><a href="#cb85-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-11"><a href="#cb85-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert silhouette object to a data frame for visualization</span></span>
<span id="cb85-12"><a href="#cb85-12" aria-hidden="true" tabindex="-1"></a>sil_df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(sil[, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]) <span class="co"># Columns: cell ID, cluster, silhouette width</span></span>
<span id="cb85-13"><a href="#cb85-13" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(sil_df) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Cluster"</span>, <span class="st">"Neighboring_Cluster"</span>, <span class="st">"Silhouette_Width"</span>)</span>
<span id="cb85-14"><a href="#cb85-14" aria-hidden="true" tabindex="-1"></a>sil_df<span class="sc">$</span>Cluster <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(sil_df<span class="sc">$</span>Cluster)</span>
<span id="cb85-15"><a href="#cb85-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb85-16"><a href="#cb85-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot silhouette widths for each cluster</span></span>
<span id="cb85-17"><a href="#cb85-17" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(sil_df, <span class="fu">aes</span>(<span class="at">x =</span> Cluster, <span class="at">y =</span> Silhouette_Width, <span class="at">fill =</span> Cluster)) <span class="sc">+</span></span>
<span id="cb85-18"><a href="#cb85-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb85-19"><a href="#cb85-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb85-20"><a href="#cb85-20" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">"Silhouette Analysis of Clusters"</span>,</span>
<span id="cb85-21"><a href="#cb85-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">"Cluster"</span>,</span>
<span id="cb85-22"><a href="#cb85-22" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">"Silhouette Width"</span></span>
<span id="cb85-23"><a href="#cb85-23" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb85-24"><a href="#cb85-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()<span class="sc">+</span><span class="fu">scale_fill_manual</span>(<span class="at">values =</span> clusterColors)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="scRNAseq-course_files/figure-html/silhouette-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="cell-cycle-analysis" class="level3">
<h3 class="anchored" data-anchor-id="cell-cycle-analysis">8.3 Cell Cycle Analysis</h3>
<p>Cell cycle phase can be a significant source of variation in scRNA-seq data, influencing gene expression patterns and clustering results. To account for this, we can use the <strong>CellCycleScoring</strong> function in Seurat to assign a cell cycle phase to each cell based on the expression of cell cycle-related genes. This information can then be used to adjust for cell cycle effects in downstream analyses.</p>
<p>For each cell, CellCycleScoring calculates the S.Score and G2M.Score by averaging the expression of predefined S-phase and G2/M-phase marker genes, respectively. These scores quantify the activity of phase-specific genes, providing a basis for classification. A cell is assigned to the S-phase if the S.Score exceeds the G2M.Score, to the G2/M-phase if the G2M.Score is higher, and to the G1-phase if both scores are low, indicating quiescence or non-dividing status. This classification captures cell cycle progression with a straightforward scoring comparison.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load predefined cell cycle gene sets from Seurat</span></span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>s.genes <span class="ot">&lt;-</span> Seurat<span class="sc">::</span>cc.genes.updated<span class="fl">.2019</span><span class="sc">$</span>s.genes  <span class="co"># Genes associated with S-phase</span></span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>g2m.genes <span class="ot">&lt;-</span> Seurat<span class="sc">::</span>cc.genes.updated<span class="fl">.2019</span><span class="sc">$</span>g2m.genes  <span class="co"># Genes associated with G2/M-phase</span></span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-5"><a href="#cb86-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Perform cell cycle scoring</span></span>
<span id="cb86-6"><a href="#cb86-6" aria-hidden="true" tabindex="-1"></a>da_integrated <span class="ot">&lt;-</span> <span class="fu">CellCycleScoring</span>(</span>
<span id="cb86-7"><a href="#cb86-7" aria-hidden="true" tabindex="-1"></a>  da_integrated,                           <span class="co"># Seurat object to classify</span></span>
<span id="cb86-8"><a href="#cb86-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">s.features =</span> s.genes,              <span class="co"># Gene set for S-phase</span></span>
<span id="cb86-9"><a href="#cb86-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">g2m.features =</span> g2m.genes,          <span class="co"># Gene set for G2/M-phase</span></span>
<span id="cb86-10"><a href="#cb86-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">set.ident =</span> <span class="cn">TRUE</span>                   <span class="co"># Update cell identities based on the predicted phase</span></span>
<span id="cb86-11"><a href="#cb86-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb86-12"><a href="#cb86-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb86-13"><a href="#cb86-13" aria-hidden="true" tabindex="-1"></a><span class="co"># View the resulting cell cycle scores</span></span>
<span id="cb86-14"><a href="#cb86-14" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(da_integrated<span class="sc">@</span>meta.data[, <span class="fu">c</span>(<span class="st">"S.Score"</span>, <span class="st">"G2M.Score"</span>, <span class="st">"Phase"</span>)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                              S.Score     G2M.Score        Phase
day60-1_AAACCTGCACCAGCAC-1 0.06656733  0.0205103063 -0.065400410
day60-1_AAACCTGGTATTCTCT-1 0.06321079 -0.0770181599 -0.036213870
day60-1_AAACCTGTCAGTTTGG-1 0.01835282 -0.0361139882 -0.055103430
day60-1_AAACCTGTCCGCTGTT-1 0.03442699 -0.0007247744 -0.027175074
day60-1_AAACCTGTCGCAAGCC-1 0.05806855 -0.0619577344 -0.037233962
day60-1_AAACGGGAGATGTAAC-1 0.08664913 -0.0192035375 -0.001779165</code></pre>
</div>
</div>
</section>
<section id="species-classification" class="level3">
<h3 class="anchored" data-anchor-id="species-classification">8.4 Species classification</h3>
<p>In some scRNA-seq experiments, cells may be derived from multiple species, such as human-mouse xenografts. We will start by creating the seurat object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read the 10X Genomics data</span></span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a>graft.data <span class="ot">&lt;-</span> <span class="fu">Read10X</span>(<span class="at">data.dir =</span> <span class="st">"data/grafts/G1_YZ013_SD_no12/"</span>) </span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove the "grch38_" prefix from gene names for easier readability</span></span>
<span id="cb88-4"><a href="#cb88-4" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(graft.data)<span class="ot">&lt;-</span><span class="fu">gsub</span>(<span class="st">"premRNAGRCH38-"</span>,<span class="st">""</span>,<span class="fu">rownames</span>(graft.data)) </span>
<span id="cb88-5"><a href="#cb88-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a Seurat object with the count data</span></span>
<span id="cb88-6"><a href="#cb88-6" aria-hidden="true" tabindex="-1"></a>graft <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> graft.data, </span>
<span id="cb88-7"><a href="#cb88-7" aria-hidden="true" tabindex="-1"></a>                            <span class="at">min.cells =</span> <span class="dv">3</span>, <span class="at">min.features =</span> <span class="dv">200</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Feature names cannot have underscores ('_'), replacing with dashes
('-')</code></pre>
</div>
</div>
<p>Next step is to use the fraction of reads mapping to rat and human genome to identify cell types.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the total counts for Rat-specific features</span></span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract features matching the Rat pattern ("premRNARnor6--") from the "counts" layer and sum them row-wise</span></span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a>ratGenes <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">"^premRNARnor6--"</span>, <span class="fu">rownames</span>(graft),<span class="at">value =</span> <span class="cn">TRUE</span>)</span>
<span id="cb90-5"><a href="#cb90-5" aria-hidden="true" tabindex="-1"></a>graft<span class="sc">$</span>RatCounts <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(<span class="fu">FetchData</span>(graft, </span>
<span id="cb90-6"><a href="#cb90-6" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">vars =</span> ratGenes, <span class="at">layer =</span> <span class="st">"counts"</span>))</span>
<span id="cb90-7"><a href="#cb90-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-8"><a href="#cb90-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the total counts for Human-specific features</span></span>
<span id="cb90-9"><a href="#cb90-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract features NOT matching the Rat pattern ("premRNARnor6--") from the "counts" layer and sum them row-wise</span></span>
<span id="cb90-10"><a href="#cb90-10" aria-hidden="true" tabindex="-1"></a>humanGenes <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">"^premRNARnor6--"</span>,</span>
<span id="cb90-11"><a href="#cb90-11" aria-hidden="true" tabindex="-1"></a>                          <span class="fu">rownames</span>(graft),</span>
<span id="cb90-12"><a href="#cb90-12" aria-hidden="true" tabindex="-1"></a>                          <span class="at">invert =</span> <span class="cn">TRUE</span>,</span>
<span id="cb90-13"><a href="#cb90-13" aria-hidden="true" tabindex="-1"></a>                          <span class="at">value =</span> <span class="cn">TRUE</span>)</span>
<span id="cb90-14"><a href="#cb90-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-15"><a href="#cb90-15" aria-hidden="true" tabindex="-1"></a>graft<span class="sc">$</span>HumanCounts <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(<span class="fu">FetchData</span>(graft, <span class="at">vars =</span> humanGenes, <span class="at">layer =</span> <span class="st">"counts"</span>))</span>
<span id="cb90-16"><a href="#cb90-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-17"><a href="#cb90-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute the percentage of counts attributed to Rat-specific features</span></span>
<span id="cb90-18"><a href="#cb90-18" aria-hidden="true" tabindex="-1"></a>graft<span class="sc">$</span>PercentRat <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(graft, <span class="at">features =</span> ratGenes)</span>
<span id="cb90-19"><a href="#cb90-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-20"><a href="#cb90-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute the percentage of counts attributed to Human-specific features</span></span>
<span id="cb90-21"><a href="#cb90-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Use the set of features that do not match the Rat-specific pattern</span></span>
<span id="cb90-22"><a href="#cb90-22" aria-hidden="true" tabindex="-1"></a>graft<span class="sc">$</span>PercentHuman <span class="ot">&lt;-</span> <span class="fu">PercentageFeatureSet</span>(graft,</span>
<span id="cb90-23"><a href="#cb90-23" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">features =</span> humanGenes)</span>
<span id="cb90-24"><a href="#cb90-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-25"><a href="#cb90-25" aria-hidden="true" tabindex="-1"></a><span class="co"># Classify cells into species based on the percentage of Rat-specific features</span></span>
<span id="cb90-26"><a href="#cb90-26" aria-hidden="true" tabindex="-1"></a>graft<span class="sc">$</span>Species <span class="ot">&lt;-</span> <span class="fu">case_when</span>(</span>
<span id="cb90-27"><a href="#cb90-27" aria-hidden="true" tabindex="-1"></a>  graft<span class="sc">$</span>PercentRat <span class="sc">&lt;</span> <span class="dv">20</span> <span class="sc">~</span> <span class="st">"Human"</span>,   <span class="co"># Less than 20% Rat features indicates a Human cell</span></span>
<span id="cb90-28"><a href="#cb90-28" aria-hidden="true" tabindex="-1"></a>  graft<span class="sc">$</span>PercentRat <span class="sc">&gt;</span> <span class="dv">80</span> <span class="sc">~</span> <span class="st">"Rat"</span>,     <span class="co"># More than 80% Rat features indicates a Rat cell</span></span>
<span id="cb90-29"><a href="#cb90-29" aria-hidden="true" tabindex="-1"></a>  <span class="cn">TRUE</span> <span class="sc">~</span> <span class="st">"Unknown"</span>                   <span class="co"># Anything in between is classified as "Unknown"</span></span>
<span id="cb90-30"><a href="#cb90-30" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb90-31"><a href="#cb90-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb90-32"><a href="#cb90-32" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate a scatter plot showing Rat counts vs. Human counts</span></span>
<span id="cb90-33"><a href="#cb90-33" aria-hidden="true" tabindex="-1"></a><span class="co"># Group the points by the classified species</span></span>
<span id="cb90-34"><a href="#cb90-34" aria-hidden="true" tabindex="-1"></a><span class="fu">FeatureScatter</span>(graft, <span class="at">feature1 =</span> <span class="st">"RatCounts"</span>, <span class="at">feature2 =</span> <span class="st">"HumanCounts"</span>, <span class="at">group.by =</span> <span class="st">"Species"</span>) <span class="sc">+</span></span>
<span id="cb90-35"><a href="#cb90-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Species Classification"</span>) <span class="sc">+</span> <span class="co"># Add a title to the plot</span></span>
<span id="cb90-36"><a href="#cb90-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Rat Counts"</span>) <span class="sc">+</span>                <span class="co"># Label the x-axis</span></span>
<span id="cb90-37"><a href="#cb90-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Human Counts"</span>)                <span class="co"># Label the y-axis</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="scRNAseq-course_files/figure-html/create_multi_species_seu2-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a subset of the Seurat object containing only cells classified as "Human"</span></span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>graft.human <span class="ot">&lt;-</span> <span class="fu">subset</span>(graft, <span class="at">subset =</span> Species <span class="sc">==</span> <span class="st">"Human"</span>)</span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a><span class="co"># And you can start with QC and so on... </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>